### PR TITLE
enhancement(vector sink): Add support for TLS

### LINKED ISSUE #1993: Reorganize and centralize TLS implementations
There are many pieces of TLS implementations sprinkled throughout the sinks and sources. In particular, any sink or source that can handle TLS communication (or HTTPS) needs to be aware of this and have at least some difference in the control flow for TLS versus non-TLS connections. This should be formalized and centralized into `src/tls.rs` as a set of `MaybeTls` types and trait implementations that can operate in both modes. This will reduce the complexity of sinks and sources that may handle both types of connections.

### LINKED ISSUE #2024: Vector sink does not support TLS
In #1892 TLS support was added to the Vector source! This is awesome, but the Vector sink does not yet support TLS.

Fixing this issue means adding a TLS option to the Vector sink and making sure it works.

The options should match alongside https://vector.dev/docs/reference/sources/vector/ where possible.

### PR DESCRIPTION
**Pull request by bruceg**
Closes #2024

### CODE REVIEW

File: src/sources/vector.rs [hunk 1]
```diff
@@ -25,12 +25,13 @@ fn default_shutdown_timeout_secs() -> u64 {
     30
 }
 
+#[cfg(test)]
```
Comment (Hoverbear) [inline src/sources/vector.rs]: This seems like it might have slipped in, is this intentional?

Comment (bruceg) [reply src/sources/vector.rs]: Very intentional. The tests for both the sink and source are in this file, so this is required to test TLS support.

**Review by Hoverbear — APPROVED** [review]
Rest looks great!

File: src/sources/vector.rs (hunk 1 above)
Comment (Hoverbear) [reply src/sources/vector.rs]: Ok!

**Review by binarylogic — APPROVED** [review]
:shipit:

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**